econml/drlearner.py

```diff
@@ -558,6 +562,46 @@ def cate_feature_names(self, input_feature_names=None):
         else:
             raise AttributeError("Featurizer does not have a method: get_feature_names!")
 
+    def shap_values(self, X, *, feature_names=None, treatment_names=None, output_names=None):
+        if self.featurizer is not None:
+            F = self.featurizer.transform(X)
+        else:
+            F = X
+        feature_names = self.cate_feature_names(feature_names)
+
+        if self._multitask_model_final:
+            d_t = self._d_t[0] if self._d_t else 1
+            if treatment_names is None:
+                treatment_names = [f"T{i}" for i in range(d_t)]
+            if output_names is None:
+                output_names = ["Y0"]
+
+            # define masker by using entire dataset, otherwise Explainer will only sample 100 obs by default.
+            background = shap.maskers.Independent(F, max_samples=F.shape[0])
+            shap_outs = defaultdict(dict)
+            try:
+                explainer = shap.Explainer(self.multitask_model_cate, background,
+                                           feature_names=feature_names)
+            except Exception as e:
+                print("Final model can't be parsed, explain const_marginal_effect() instead!")
+                return _shap_explain_cme(self.const_marginal_effect, F, 1, d_t, feature_names, treatment_names,
+                                         output_names)
+
+            shap_out = explainer(F)
+            if d_t > 1:
+                for i in range(d_t):
+                    shap_out_copy = deepcopy(shap_out)
+                    shap_out_copy.base_values = shap_out_copy.base_values[..., i]
+                    shap_out_copy.values = shap_out_copy.values[..., i]
+                    shap_outs[output_names[0]][treatment_names[i]] = shap_out_copy
+            else:
+                shap_outs[output_names[0]][treatment_names[0]] = shap_out
+            return shap_outs
+        else:
+            return super()._shap_values(super().model_final.models_cate, F, feature_names=feature_names,
+                                        treatment_names=treatment_names, output_names=output_names)
```

Comment:
A lot of this logic looks extremely similar to LinearCateEstimator's `_shap_values` and also LinearModelFinalCateEstimatorMixin's `shap_values`.  Is it possible to extract the common logic so that it's all in just one place and making changes will be easier in the future?

Reply:
Yeah I thought about that.  the only part repeats multiple times is 
```
        d_t = self._d_t[0] if self._d_t else 1
        d_y = self._d_y[0] if self._d_y else 1

        if treatment_names is None:
            treatment_names = [f"T{i}" for i in range(d_t)]
        if output_names is None:
            output_names = [f"Y{i}" for i in range(d_y)]
```
Any idea on how to avoid duplicates for this?

For the loop to extract shap values of each T and each Y, they look similar but they are different given the structure of final model. `drlearner` for mulitple T is one single multi-output model, `dml` for multiple T should filter cate each given T, and the for loop under `_shap_values` for multiple T is one model per T, all those slice the `shap_values` in different way.

Reply:
If the details vary because the indexing is subtly different then maybe there's not much we can do to simplify the logic, in which case don't worry about it.